src/libponyc/type/lookup.c

```diff
@@ -178,6 +216,10 @@ static deferred_reification_t* lookup_nominal(pass_opt_t* opt, ast_t* from,
   }
 
   ast_t* typeargs = ast_childidx(type, 2);
+
+  if(ast_id(find) == TK_FUN && ast_id(ast_child(find)) == TK_BOX)
+    orig = downcast_iso_trn_receiver_to_ref(orig);
+
   return deferred_reify_new(find, typeparams, typeargs, orig);
 }
```

Comment:
This needs an `ast_free` call with `orig` if `downcast_iso_trn_receiver_to_ref` did a copy, as `deferred_reify_new` also does a copy. Another possibility would be to add a new version of `deferred_reify_new` that doesn't copy the AST for the `this` type and use it in that case.